gcovr/gcov_parser.py

```diff
@@ -954,86 +967,166 @@ def line_after_start(self, lineno: int, start: str, start_lineno: str) -> None:
+
+
 def _find_excluded_ranges(
     lines: List[Tuple[int, str]],
     *,
     warnings: _ExclusionRangeWarnings,
-    exclude_lines_by_pattern: Optional[str] = None,
+    exclude_lines_by_custom_pattern: Optional[str] = None,
+    exclude_branches_by_custom_pattern: Optional[str] = None,
     exclude_pattern_prefix: str,
-) -> Callable[[int], bool]:
+) -> Tuple[Callable[[int], bool]]:
     """
-    Scan through all lines to find line ranges covered by exclusion markers.
+    Scan through all lines to find line ranges and branch ranges covered by exclusion markers.
 
     Example:
-    >>> lines = [(11, '//PREFIX_EXCL_LINE'), (13, '//IGNORE'), (15, '//PREFIX_EXCL_START'), (18, '//PREFIX_EXCL_STOP')]
-    >>> exclude = _find_excluded_ranges(
-    ...     lines, warnings=..., exclude_lines_by_pattern = '.*IGNORE', exclude_pattern_prefix='PREFIX')
-    >>> [lineno for lineno in range(20) if exclude(lineno)]
+    >>> lines = [(11, '//PREFIX_EXCL_LINE'), (13, '//IGNORE_LINE'), (15, '//PREFIX_EXCL_START'), (18, '//PREFIX_EXCL_STOP'),
+    ...     (21, '//PREFIX_EXCL_BR_LINE'), (23, '//IGNORE_BR'), (25, '//PREFIX_EXCL_BR_START'), (28, '//PREFIX_EXCL_BR_STOP')]
+    >>> exclude_line, exclude_branch = _find_excluded_ranges(
+    ...     lines, warnings=..., exclude_lines_by_custom_pattern = '.*IGNORE_LINE',
+    ...     exclude_branches_by_custom_pattern = '.*IGNORE_BR', exclude_pattern_prefix='PREFIX')
+    >>> [lineno for lineno in range(30) if exclude_line(lineno)]
     [11, 13, 15, 16, 17]
+    >>> [lineno for lineno in range(30) if exclude_branch(lineno)]
+    [21, 23, 25, 26, 27]
     """
-    exclude_lines_by_pattern_regex = None
-    if exclude_lines_by_pattern:
-        exclude_lines_by_pattern_regex = re.compile(exclude_lines_by_pattern)
+    exclude_lines_by_custom_pattern_regex = None
+    if exclude_lines_by_custom_pattern:
+        exclude_lines_by_custom_pattern_regex = re.compile(
+            exclude_lines_by_custom_pattern
+        )
 
-    # possibly overlapping half-open ranges of lines that are excluded
-    exclude_line_ranges: List[Tuple[int, int]] = []
+    exclude_branches_by_custom_pattern_regex = None
+    if exclude_branches_by_custom_pattern:
+        exclude_branches_by_custom_pattern_regex = re.compile(
+            exclude_branches_by_custom_pattern
+        )
 
-    exclusion_stack = []
+    # possibly overlapping half-open ranges that are excluded
+    exclude_line_ranges: List[Tuple[int, int]] = []
+    exclude_branch_ranges: List[Tuple[int, int]] = []
+
+    exclusion_stack_line = []
+    exclusion_stack_branch = []
+
+    excl_line_pattern = re.compile(
+        "("
+        + exclude_pattern_prefix
+        + ")"
+        + _EXCLUDE_FLAG
+        + _EXCLUDE_LINE_WORD
+        + _EXCLUDE_PATTERN_POSTFIX
+    )
+    excl_branch_pattern = re.compile(
+        "("
+        + exclude_pattern_prefix
+        + ")"
+        + _EXCLUDE_FLAG
+        + _EXCLUDE_BRANCH_WORD
+        + _EXCLUDE_PATTERN_POSTFIX
+    )
```

Comment:
This operation is only needed if `_EXCLUDE_FLAG` is found in code (performance issue). You can set the values to `None` here and init the regex on first entry of if block in line 1083.

Reply:
Indeed, I thought it was already improving the performance if it was defined only once (it was for each exclude line before), but it's even better if it's declared only when it's needed!

Fixed